xdsl/backend/riscv/prologue_epilogue_insertion.py

```diff
@@ -0,0 +1,92 @@
+from xdsl.builder import Builder
+from xdsl.dialects import builtin, riscv, riscv_func
+from xdsl.dialects.riscv import (
+    FloatRegisterType,
+    IntRegisterType,
+    Registers,
+    RISCVRegisterType,
+)
+from xdsl.ir import MLContext
+from xdsl.passes import ModulePass
+
+
+class PrologueEpilogueInsertion(ModulePass):
+    """
+    Pass inserting a prologue and epilogue according to the RISC-V ABI.
+    The prologues and epilogues are responsible for saving any callee-preserved
+    registers.
+    In RISC-V these are 's0' to 's11' and 'fs0' to `fs11'.
+    The stack pointer 'sp' must also be restored to its original value.
+
+    This pass should be run late in the pipeline after register allocation.
+    It does not itself require register allocation nor invalidate the result of the
+    register allocator.
+    """
+
+    name = "riscv-prologue-epilogue-insertion"
+
+    def _process_function(self, func: riscv_func.FuncOp) -> None:
+        # Find all callee-preserved registers that are clobbered. We define clobbered
+        # as it being the result of some operation and therefore written to.
+        used_callee_preserved_registers = dict[
+            IntRegisterType | FloatRegisterType, None
+        ]()
+        for op in func.walk():
+            # Special cases ops that do not actually write to the register.
+            if isinstance(op, riscv.GetRegisterOp | riscv.GetFloatRegisterOp):
+                continue
+
+            for res in op.results:
+                if res.type in Registers.S or res.type in Registers.FS:
+                    used_callee_preserved_registers[res.type] = None
+
+        # Note: This assumes RV32D aka 4 byte integer registers, 8 byte FP registers.
+        def get_register_size(r: RISCVRegisterType):
+            if isinstance(r, IntRegisterType):
+                return 4
+            return 8
```

Comment:
Could you please pass these as xlen and flen? Ideally we'd want these as parameters on the pass, and to have a way to share the logic between passes.

Reply:
By that I mean that in the future, we would find a way to pass xlen and flen only once ƒor all the passes, but for now it would be good to at least state them explicitly

Reply:
I made these a pass option now as that was actually really nice and easy